### ISSUE TITLE
Server crash after running several days

### ISSUE THREAD

**Issue by scottf-tvw** [typesense/typesense#70]
## Description

```
2019/06/09 09:04:59 793538      INFO [typesense_server_utils.cpp->run_server:69]        Starting Typesense 0.9.2
2019/06/09 09:05:00 124710      INFO [typesense_server_utils.cpp->run_server:79]        Loading collections from disk...
2019/06/09 09:56:58 147884      INFO [typesense_server_utils.cpp->run_server:86]        Finished loading collections from disk.
2019/06/09 09:56:58 150341      INFO [http_server.cpp->run:144] Typesense has started. Ready to accept requests on port 8108
2019/06/12 21:07:37 373086

***** FATAL SIGNAL RECEIVED *******
Received fatal signal: SIGABRT(6)       PID: 12575

***** SIGNAL SIGABRT(6)

******* STACKDUMP *******
        stack dump [1]  typesense-server() [0x6b0f68]
        stack dump [2]  /lib/x86_64-linux-gnu/libpthread.so.0+0x11390 [0x7fc4292c8390]
        stack dump [3]  /lib/x86_64-linux-gnu/libc.so.6gsignal+0x38 [0x7fc42880d428]
        stack dump [4]  /lib/x86_64-linux-gnu/libc.so.6abort+0x16a [0x7fc42880f02a]

        stack dump [5]  typesense-server : __gnu_cxx::__verbose_terminate_handler()+0x15d [0xb4d16d]

        stack dump [6]  typesense-server : __cxxabiv1::__terminate(void (*)())+0x6 [0xaba906]
        stack dump [7]  typesense-server() [0xaba951]
        stack dump [8]  typesense-server() [0xad74de]
        stack dump [9]  /lib/x86_64-linux-gnu/libpthread.so.0+0x76ba [0x7fc4292be6ba]
        stack dump [10]  /lib/x86_64-linux-gnu/libc.so.6clone+0x6d [0x7fc4288df41d]

Exiting after fatal event  (FATAL_SIGNAL). Fatal type:  SIGABRT
Log content flushed sucessfully to sink


g3log g3FileSink shutdown at: 21:07:37 373247
```

## Steps to reproduce

server starts from supervisor using
`typesense-server --data-dir=/invintus/typesenseData/ --api-key="xxxxxxxxx" --enable-cors --log-dir /invintus/logs`

## Expected Behavior

Not to crash dump

## Actual Behavior

Unknown why error occuring.

## Metadata

**RAM DISK allocation**: `KiB Mem : 32560808 total,  1020500 free, 26474852 used,  5065456 buff/cache`

**Typsense Version**: `0.9.2`

**OS**: `Ubuntu 16.04.5 LTS`

**Comment by kishorenc** [comment]
@scottf-tvw 

We do a `SIGABRT` when a system memory allocation call fails. From the RAM allocation metadata you have provided, it seems like there is only about 1 GB of free memory out of 32 GB. The free memory available could have spiked or tipped during the course of running, which might have resulted in a memory allocation call failing. Do you colocate Typesense with other services running on the box?

**Comment by MyPustak** [comment]
@kishorenc Do we should keep typesense server seperate from our backend server . Please suggest how we can allocate more RAM for typesense

**Comment by kishorenc** [comment]
@MyPustak You don't have to keep it separately if there is enough RAM. If you are colocating Typesense with another service whose memory requirements tend to be spiky in nature, then there could be a scenario when the overall memory is not enough during a brief period of time. In such a case, `SIGABRT` will happen. 

> Please suggest how we can allocate more RAM for typesense

There is no pre-allocation of memory. Typesense acquires more memory when it needs and grows elastically as long as there is enough system memory available. Happy to discuss further about your specific setup offline, if you wish. Email me at contact at typesense dot org.

**Comment by MyPustak** [comment]
Thanks a lot for your prompt response.

**Comment by kishorenc** [comment]
@MyPustak If there is enough memory on the machine and you are still facing this issue, please give me more details on when exactly the crash happens (e.g. during indexing or searching). If you can share the dataset with me privately, we can also run further tests on it to see if we can reproduce the behavior.

**Comment by kishorenc** [comment]
We found an edge case that triggered a crash sometimes on really long queries. I've pushed out a fix for that, and also built a new release for you to try. Can you please download and use the following release?

https://dl.typesense.org/releases/typesense-server-0.10.1-gamma-linux-amd64.tar.gz

I think that crash issue should be addressed with that. Let me know how it goes. This fix will be part of our future `0.10.1` release soon.

**Comment by kishorenc** [comment]
[Version `0.10.1`](https://github.com/typesense/typesense/releases/tag/v0.10.1) is out now and it contains a fix for the `SIGABRT` crash on long query strings. I'm going to close this issue for now. Feel free to reopen if it resurfaces again on this version.
